pkg/mcp/tests/stress/stress_test.go

```diff
@@ -0,0 +1,634 @@
+					b.SetEntry(collection, e.name, e.version.String(), e.createTime, e.labels, e.annotations, e.body)
+				case 2: // Delete a resource
+					l := resourcesByCollection[collection]
+					if len(l) == 0 {
+						continue
+					}
+					idx := rnd.Intn(len(l))
+					e := l[idx]
+					b.DeleteEntry(collection, e.name)
+					l = append(l[:idx], l[idx+1:]...)
+					resourcesByCollection[collection] = l
+				}
+			}
+
+			sortedResources := resourcesByCollection[collection]
+			sort.Slice(sortedResources, func(i, j int) bool { return sortedResources[i].name < sortedResources[j].name })
+
+			// The response's resourceVersion is a hash of the name and version of each resource in the collection.
+			h := sha256.New()
+			for _, entry := range sortedResources {
+				h.Write([]byte(entry.name))
+				h.Write([]byte(entry.version.String()))
+			}
+			version := base64.StdEncoding.EncodeToString(h.Sum(nil))
+			b.SetVersion(collection, version)
+		}
+		sn := b.Build()
+		commonDataset = append(commonDataset, sn)
+		prev = sn
+	}
+}
+
+type clientState struct {
+	// metadata indexed by collection and name
+	resourcesByCollectionName map[string]map[string]*mcp.Metadata
+
+	applied    int64
+	badVersion int64
+	acked      int64
+	nacked     int64
+	added      int64
+	removed    int64
+	inc        int64
+
+	minApplyDelay time.Duration
+	maxApplyDelay time.Duration
+	nackRate      float64
+
+	rnd *rand.Rand
+}
+
+type driver struct {
+	options options
+	rnd     *rand.Rand
+
+	clientOpts *sink.Options
+	serverOpts *source.Options
+
+	nextDataset int
+
+	ctx    context.Context
+	cancel context.CancelFunc
+
+	listener      net.Listener
+	grpcServer    *grpc.Server
+	serverAddress string
+	cache         *snapshot.Cache
+
+	limiter *rate.Limiter
+
+	clientWG sync.WaitGroup
+	clients  []*clientState
+}
+
+type options struct {
+	stepRate         time.Duration
+	numClients       int
+	iterations       int
+	minUpdateDelay   time.Duration
+	maxUpdateDelay   time.Duration
+	minApplyDelay    time.Duration
+	maxApplyDelay    time.Duration
+	nackRate         float64
+	serverIncemental bool
+	clientIncRate    float64
+	setupFn          func(d *driver)
+	serverStepFn     func(d *driver)
+}
+
+func updateSnapshot(d *driver) {
+	time.Sleep(randomTimeInRange(d.rnd, d.options.minUpdateDelay, d.options.maxUpdateDelay))
+
+	sn := commonDataset[d.nextDataset]
+
+	d.nextDataset = (d.nextDataset + 1) % len(commonDataset)
+
+	d.cache.SetSnapshot(groups.Default, sn)
+}
+
+func defaultOptions() options {
+	return options{
+		stepRate:         *stepRate,
+		numClients:       *numClients,
+		minApplyDelay:    *minApplyDelay,
+		maxApplyDelay:    *maxApplyDelay,
+		minUpdateDelay:   *minUpdateDelay,
+		maxUpdateDelay:   *maxUpdateDelay,
+		nackRate:         *nackRate,
+		iterations:       *iterations,
+		serverIncemental: true,
+		serverStepFn:     updateSnapshot,
+	}
+}
+
+func newDriver(o options) (*driver, error) {
```

Comment:
Consider having different file for options/driver and the tests themselves.

Reply:
Moved most of the setup code to stress_setup_test.go.